Sources/App/Extensions/Vapor.Request+RequestProtocol.swift

```diff
@@ -0,0 +1,14 @@
+import Foundation
+import Bot
+import Vapor
+
+extension Vapor.Request: RequestProtocol {
+    public func header(named name: String) -> String? {
+        return http.headers[name].first
+    }
+
+    public func decodeBody<T>(_ type: T.Type) -> T? where T: Decodable {
+        return try? content.decode(type).wait()
```

Comment:
I have to admit I'm hesitant about the whole "Avoid using Futures just to make the code independent of Vapor" thing. Might be premature overarchitecturing which introduce main thread blocking like that just for it — even if this shouldn't be such a big thread blocker, still not good in principle.

Reply:
the thing is that it still does the same what asyncDecode does where the problem is in `wait` called on the event loop (iirc SwiftNIO docs explicitly say that there should be no blocking calls when code is running in the event loop) 🤷‍♂ I was baffled by that myself and have a branch in Stevenson where actually replaced use of `asyncDecode` with just `decode` (we have futures everywhere already so it fits in well)
Probably to avoid that this method should return a signal or signal producer if you want to avoid depending on Futures.
See the SwiftNIO code for reference (it's part of Vapor so you can just search for it in Xcode) - `EventLoopFuture.swift:799`.

Reply:
We are now using `SignalProducer`s which start on their own scheduler, please let me know if you see any potential issue or theres'a a better way to execute a future.